### PR TITLE
Update AirNow converter

### LINKED ISSUE #1297: Write or update a converter for surface AQ observations
## Description

Write or update a ioda-converter to ingest either airnow or open AQ observations.
https://github.com/JCSDA-internal/ioda-converters/blob/develop/src/compo/airnow2ioda_nc.py

## Requirements

For pm2.5 if the model output doesn't produce pm2.5 we would need to add a VADER transform.
@jeromebarre can help with this

## Acceptance Criteria (Definition of Done)

The converter can take O3, PM2.5, NO2, SO2 (maybe CO) and data is tested in UFO

## Dependencies

TBD

### PR DESCRIPTION
**Pull request by weiwilliam**
## Description
This PR updates the converter for AirNow observations with, 
1. Add land use to variable `airqualityClassification` (0=UNKNOWN, 1=RURAL, 2=SUBURBAN, 3=URBAN AND CENTER CITY) based on a full site list provided by EPA
2. Users can define whether the site list contains the land use information by giving `--epa_list` argument
3. multi-files capability 

## Issue(s) addressed

Resolves #1297 

## Dependencies
No dependencies

## Impact

Expected impact on downstream repositories:

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have run the unit tests before creating the PR

### CODE REVIEW

**Comment by weiwilliam** [comment]
@jeromebarre and @mer-a-o Is `surfaceQualifier` good to store land use information?

File: src/compo/airnow2ioda_nc.py [hunk 1]
```diff
@@ -12,30 +12,39 @@
...
                'ioda_version': 2,
                'description': 'AIRNow data (converted from text/csv to IODA',
-               'source': 'Unknown (ftp)'}
+               'source': 'https://files.airnowtech.org/?prefix=airnow/',
+               'sourceFiles': '',
+               }
 
-iso8601_string = locationKeyList[meta_keys.index('dateTime')][2]
-epoch = datetime.fromisoformat(iso8601_string[14:-1])
+#iso8601_string = locationKeyList[meta_keys.index('dateTime')][2]
```
Comment (mer-a-o) [inline src/compo/airnow2ioda_nc.py]: remove these lines if they are not needed.

File: src/compo/airnow2ioda_nc.py [hunk 2]
```diff
@@ -133,63 +150,85 @@ def add_data(infile, sitefile):
...
         help="path of AIRNow text input file",
-        type=str, required=True)
+        type=str, nargs='+', required=True)
     required.add_argument(
         '-s', '--sitefile',
         help="path of AIRNow site list file",
         type=str, required=True)
+    required.add_argument(
+        '--fulllist',
+        help="define this if sitefile is a full list",
```
Comment (mer-a-o) [inline src/compo/airnow2ioda_nc.py]: What does full list mean? Can you make it a little more clear?

**Review by mer-a-o — CHANGES_REQUESTED** [review]
The test is failing for me:
```
66: Traceback (most recent call last):
66:   File "/discover/nobackup/mabdiosk/jedi-bundle/build-gnu-spack-1.7.0-mil/bin/airnow2ioda_nc.py", line 203, in <module>
66:     time_offset = round((dt - epoch).total_seconds())
66: TypeError: can't subtract offset-naive and offset-aware datetimes
```
Can you add an option to take the assimilation window from the user and convert only within that window? Similar to what @jeromebarre added here: https://github.com/JCSDA-internal/ioda-converters/blob/develop/src/compo/mopitt_co_nc2ioda.py#L152

File: src/compo/airnow2ioda_nc.py (hunk 2 above)
Comment (weiwilliam) [reply src/compo/airnow2ioda_nc.py]: I will update it to `--epa_list` and modify the description to "define this if the sitefile contains land use information from EPA".

**Comment by mer-a-o** [comment]
@weiwilliam I rebuilt the code and the error went away. But I'm getting a different error:
```
66: Processing infile= testinput/airnow_2020081306.dat
66: DIFFER : NUMBER OF GLOBAL ATTRIBUTES : 7 <> 6
66: DIFFER : VALUES OF GLOBAL ATTRIBUTE : source : "https://files.airnowtech.org/?prefix=airnow/" <> "Unknown (ftp)"
66: DIFFER : NAME OF GLOBAL ATTRIBUTE : sourceFiles : GLOBAL ATTRIBUTE DOESN'T EXIST IN "testoutput/airnow_2020081306.nc"
66: DIFFER : VARIABLE : carbonmonoxideSurface : DOES NOT EXIST IN "testoutput/airnow_2020081306.nc"
66: DIFFER : VARIABLE : nitrogendioxideSurface : DOES NOT EXIST IN "testoutput/airnow_2020081306.nc"
66: DIFFER : VARIABLE : sulfurdioxideSurface : DOES NOT EXIST IN "testoutput/airnow_2020081306.nc"
66: DIFFER : VARIABLE : carbonmonoxideSurface : DOES NOT EXIST IN "testoutput/airnow_2020081306.nc"
66: DIFFER : VARIABLE : nitrogendioxideSurface : DOES NOT EXIST IN "testoutput/airnow_2020081306.nc"
66: DIFFER : VARIABLE : sulfurdioxideSurface : DOES NOT EXIST IN "testoutput/airnow_2020081306.nc"
66: DIFFER : VARIABLE : carbonmonoxideSurface : DOES NOT EXIST IN "testoutput/airnow_2020081306.nc"
66: DIFFER : VARIABLE : nitrogendioxideSurface : DOES NOT EXIST IN "testoutput/airnow_2020081306.nc"
66: DIFFER : VARIABLE : sulfurdioxideSurface : DOES NOT EXIST IN "testoutput/airnow_2020081306.nc"
1/1 Test #66: test_iodaconv_airnow .............***Failed    1.51 sec
```

**Comment by weiwilliam** [comment]
@mer-a-o Yes, the testoutput is not updated, because I would like to know whether `surfaceQualifier` is good to use. The current usages are 0=land, 1=sea, and 2=seaice.

**Comment by mer-a-o** [comment]
How about `stationType`:  https://github.com/JCSDA-internal/ioda/blob/develop/share/ioda/yaml/validation/ObsSpace.yaml#L1075?

**Comment by weiwilliam** [comment]
I also considered that, but the description said "type of station such as ASOS, AWOS".
Then it would be "AirNow" in this case.
I use "surfaceQualifier" because there is another variable "earthSurfaceType" noted with "better for future".

@BenjaminRuston Can you suggest a variable that fits "land use" information better, such as "rural" and "urban"? Thanks.

**Comment by jeromebarre** [comment]
replace `surfaceQualifier` by `airqualityClassification`
i.e. https://www.eea.europa.eu/en/topics/in-depth/air-pollution/monitoring-station-classifications-and-criteria

**Review by jeromebarre — CHANGES_REQUESTED** [review]
Looks good to me I haven't had time to test yet. Please address the comment of the classification variable above.

**Comment by weiwilliam** [comment]
@jeromebarre Thank you! I will update it and address Maryam's comments.

**Comment by weiwilliam** [comment]
We will need to add the following variables into ObsSpace.yaml for validation:
```
MetaData/airqualityClassification
[ObsValue, ObsError, PreQC]/particulatematter2p5Surface
[ObsValue, ObsError, PreQC]/ozoneSurface
[ObsValue, ObsError, PreQC]/carbonmonoxideSurface
[ObsValue, ObsError, PreQC]/nitrogendioxideSurface
[ObsValue, ObsError, PreQC]/sulfurdioxideSurface
```

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -1358,7 +1360,19 @@ ecbuild_add_test( TARGET  test_${PROJECT_NAME}_airnow
                           -i testinput/airnow_2020081306.dat
                           -s testinput/airnow_sites.dat
                           -o testrun/airnow_2020081306.nc"
```
Comment (mer-a-o) [inline test/CMakeLists.txt:1362]: @weiwilliam you are missing the last line of `ecbuild_add_test` command here.

Comment (mer-a-o) [reply test/CMakeLists.txt:1362]: Are you going to update `testoutput/airnow_2020081306.nc` ?

Comment (weiwilliam) [reply test/CMakeLists.txt:1362]: No, there is another new test `test_${PROJECT_NAME}_airnow_epalist` below, which use the same input but different output.

Comment (weiwilliam) [reply test/CMakeLists.txt:1362]: @mer-a-o thank you, I see the missing line.

**Comment by mer-a-o** [comment]
@weiwilliam Do the airnow tests pass for you? Do you need to update the reference files?

**Comment by weiwilliam** [comment]
@mer-a-o I put the csv file in the wrong place, and I will update the variables in original testoutput.

**Review by mer-a-o — APPROVED** [review]

**Comment by mer-a-o** [comment]
@weiwilliam I'm looking at `testrun/airnow_2020081306.nc` why all the values are 0.1 for all the measured species?

**Comment by weiwilliam** [comment]
@mer-a-o let me check.

**Comment by weiwilliam** [comment]
@mer-a-o were you checking ObsError? ObsError is filled with 0.1 at all data points.

**Comment by mer-a-o** [comment]
@weiwilliam Yes!! Sorry for the false alarm.

**Comment by jeromebarre** [comment]
> @mer-a-o were you checking ObsError? ObsError is filled with 0.1 at all data points.

This is because Airnow doesn't provide the obs error. Should we set the error to a fraction of the measured quantity? The fraction should change for each species and class I think? In the future, if we want to assimilate air now it would be good to do a little bit of research to see if in the literature there is a way to estimate the error for DA and implement it in the converter. If this doesn't exist this is a little project that we should carry forward. This is important for regional AQ DA.

**Review by jeromebarre — APPROVED** [review]

File: src/compo/airnow2ioda_nc.py [hunk 4]
```diff
@@ -1,7 +1,14 @@
...
+#
+# (C) Copyright 2024 UCAR
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+#
 # Read airnow text data file and convert to IODA netcdf
-import os
+
+import os, sys
```
Comment (BenjaminRuston) [inline src/compo/airnow2ioda_nc.py]:
```suggestion
import os
```

i do not see this being used

**Comment by BenjaminRuston** [comment]
not particularly satisfied with the state of `surfaceQualifier` as in the ioda validation yaml, and this PR is perfectly valid but the values are ambiguous as to what they refer to:

https://github.com/JCSDA-internal/ioda/blob/d49ed17e115f5d7ff3310cd1dfe4798c10aa03db/share/ioda/yaml/validation/ObsSpace.yaml#L1274C1-L1274C80

```
  - Variable: [ "surfaceQualifier", "surface_type" ]  # 0=land, 1=sea, 2=seaice
  ```

**Comment by weiwilliam** [comment]
@BenjaminRuston Sorry, I forgot to update the PR description at the top. We are using `airqualityClassification` now.

**Comment by BenjaminRuston** [comment]
> @BenjaminRuston Sorry, I forgot to update the PR description at the top. We are using `airqualityClassification` now.

yes seeing that... the `airqualityClassification` is not in the IODA standards we need to add this and am changing it to `airQualityClassification`

will update this PR, please note this change and lets use this moving forward

**Comment by BenjaminRuston** [comment]
@weiwilliam updated the name please review and let me know if this work

we will need to update:
share/ioda/yaml/validation/ObsSpace.yaml

please go ahead and put a PR into IODA to add this new Qualifier and we should have folks review across agencies

**Review by BenjaminRuston — APPROVED** [review]
@weiwilliam this seems fine, there are two `airnow_sites` files now... one which has the `epalist` and requires the argument it appears.  We could potentially also make this option irrelevant if the `airnow_sites` file is the driver behind this true or false of the epalist

**Comment by BenjaminRuston** [comment]
@weiwilliam we should talk there's likely a JEDI integration test expecting this psuedo kind-of like a camel case name `airqualityClassification` ... but this is another thing would like to fix as:

image


and though I completely understand that this is an "air-quality" Classification.... camelCase is agnostic to the words and would like to just follow and not redefine *camel case* with new entries

**Pull request merged**